File: Coding_Problems_Java/Reorder List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        if(head == null || head.next == null) return;
        Deque<ListNode> Q = new LinkedList<>();
        ListNode temp = head;

        while(temp != null){
            Q.offerLast(temp);
            temp = temp.next;
        }
        head = Q.pollFirst();

        while(head != null){
            head.next = Q.pollLast();
            head = head.next;
            if (!Q.isEmpty()){
                head.next = Q.pollFirst();
                head = head.next;
            }
        }

    }
}

File: Coding Problems - Python/Linked List/Reorder List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
from collections import deque
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        
        Q = deque()
        if not head or not head.next:
            return 

        temp = head
        while temp:
            Q.append(temp)
            temp = temp.next
        
        head = Q.popleft()

        while Q:
            head.next = Q.pop()
            head = head.next
            if Q:
                head.next = Q.popleft()
                head = head.next
        head.next = None
